src/config.rs

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright 2022 Two Sigma Open Source, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//! Configuration for nsncd.
+
+use std::time::Duration;
+use std::{collections::BTreeMap, env};
+
+use anyhow::{Context, Result};
+
+use super::protocol::RequestType;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Config {
+    pub worker_count: usize,
+    pub handoff_timeout: Duration,
+    pub ignored_request_types: [bool; 32],
+}
+
+/// Mapping from nsswitch.conf "database" name to the request types related to
+/// that database.
+const OPS_BY_DATABASE: &[(&str, &[RequestType])] = &[
+    (
+        "group",
+        &[RequestType::GETGRBYNAME, RequestType::GETGRBYGID],
+    ),
+    (
+        "hosts",
+        &[
+            RequestType::GETHOSTBYADDR,
+            RequestType::GETHOSTBYADDRv6,
+            RequestType::GETHOSTBYNAME,
+            RequestType::GETHOSTBYNAMEv6,
+            RequestType::GETAI,
+        ],
+    ),
+    ("initgroups", &[RequestType::INITGROUPS]),
+    (
+        "netgroup",
+        &[RequestType::GETNETGRENT, RequestType::INNETGR],
+    ),
+    (
+        "passwd",
+        &[RequestType::GETPWBYNAME, RequestType::GETPWBYUID],
+    ),
+    (
+        "services",
+        &[RequestType::GETSERVBYNAME, RequestType::GETSERVBYPORT],
+    ),
+];
+
+impl Config {
+    /// Parse config out of the environment.
+    ///
+    /// There are two integer variables we pay attention to:
+    /// `NSNCD_WORKER_COUNT` and `NSNCD_HANDOFF_TIMEOUT`. Both must be positive
+    /// (non-zero).
+    ///
+    /// We also pay attention to variables `NSNCD_IGNORE_<DATABASE>` where
+    /// `<DATABASE>` is one of the database names from `nsswitch.conf(5)`,
+    /// capitalized:
+    ///
+    /// - NSNCD_IGNORE_GROUP
+    /// - NSNCD_IGNORE_HOSTS
+    /// - NSNCD_IGNORE_INITGROUPS
+    /// - NSNCD_IGNORE_NETGROUP
+    /// - NSNCD_IGNORE_PASSWD
+    /// - NSNCD_IGNORE_SERVICES
+    ///
+    /// These variables must be either `true` or `false`. The default is
+    /// `false` (don't ignore any requests). If one of these variables is set
+    /// to true, `nsncd` will not respond to the requests related to that
+    /// database.
+    ///
+    /// Some request types may be ignored by the implementation (e.g. the ones
+    /// that request a file descriptor pointing into internal cache
+    /// structures).
```

Comment:
Let's add the config env vars to the README as well.

Reply:
done